feat(authenticator): Add TextEditingController support to form fields

angelorodem opened the pull request:
This PR adds `TextEditingController` support to Amplify Authenticator form fields, enabling developers to programmatically control text input fields during authentication flows. It also introduces per-field visibility and enabled-state overrides for Sign Up fields so apps can keep legacy attributes off-screen or block user edits while still syncing values via controllers.

cc: @tyllark 

### What was changed
- Added new `AuthenticatorTextFieldController` class that wraps Flutter's `TextEditingController` with identical semantics.
- Added optional `controller` parameter to all text-based form fields across all authentication steps:
  - SignIn (username, password)
  - SignUp (username, email, password, phone number, address, and all custom attributes)
  - ConfirmSignIn (verification code, new password, custom auth)
  - ConfirmSignUp (username, verification code)
  - ResetPassword (username, verification code, new password)
  - VerifyUser (verification code)
  - EmailSetup (email)
- Added `visible` (default `true`) and `enabled` overrides to every `SignUpFormField` factory so applications can hide fields while keeping controller updates, or disable manual editing for attributes filled from background processes.
- Implemented bidirectional synchronization between controllers and internal authenticator state, including post-frame buffering to avoid setState-during-build issues.
- Added proper lifecycle management with post-frame callbacks to ensure correct sync timing and included diagnostics exposure for `controller`, `enabled`, and `visible` values.

### Why was this change made
Addresses the need for programmatic control over authenticator form fields to enable use cases such as:
- Pre-populating fields with data from external sources (GPS coordinates, API responses)
- Auto-filling verification codes from SMS or other sources
- Integration with third-party form validation libraries
- Dynamic form manipulation based on business logic
- Keeping legacy/system attributes hidden from end users while still providing values
- Disabling fields that should not accept manual input when values are provided automatically

### How was this change tested
- Added comprehensive unit/widget test suite with 13 test cases covering:
  - Bidirectional sync between controller and state
  - Multiple controllers on different fields
  - Rapid text updates and special key handling
  - Standard `TextEditingController` compatibility
  - Initial value population from controllers
  - Controller disposal and lifecycle management
  - Hidden and disabled Sign Up fields continuing to sync via controllers without user interaction
- All existing tests pass (301 tests)
- Created example app (`authenticator_with_controllers.dart`) demonstrating real-world usage patterns

## Related Issues
Fixes #6161

## Migration Guide
This is a non-breaking change. Existing code continues to work without modifications. To use the new controller functionality:

```dart
// Create a controller
final usernameController = AuthenticatorTextFieldController(text: 'initial_value');

// Pass it to a form field
Authenticator(
  signUpForm: SignUpForm.custom(
    fields: [
      SignUpFormField.username(
        controller: usernameController,
        enabled: false, // optional
        visible: false, // optional
      ),
      // ... other fields
    ],
  ),
  // ...
);

// Update programmatically
usernameController.text = 'new_value';

// Dispose when done
usernameController.dispose();
```

Standard `TextEditingController` is also supported for flexibility.

## Screenshots/Videos
https://github.com/user-attachments/assets/2d3d730f-9555-4f4c-8f80-cf38582ff8a3

---

**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

angelorodem commented:
@tyllark could you please review?

angelorodem commented:
@tyllark @ekjotmultani can you guys please review?

tyllark commented on packages/authenticator/amplify_authenticator/CHANGELOG.md:
```diff
@@ -1,3 +1,15 @@
+  - Bidirectional sync between controller and internal state
+  - Compatible with standard `TextEditingController` for flexibility
+- feat(authenticator): Allow SignUpFormField inputs to be disabled or hidden so apps can prefill values programmatically or keep legacy attributes off-screen
+
```
Please remove this change as we don't know the version number this will be released in. We have a package `aft` for preparing releases and it will generate the CHANGELOG updates from commit messages.

tyllark commented on packages/authenticator/amplify_authenticator/lib/src/widgets/form_field.dart:
```diff
@@ -96,6 +98,20 @@ abstract class AuthenticatorFormField<
+  /// Whether the field can receive manual input.
+  ///
+  /// When `null`, the widget decides its enabled state.
+  final bool? enabledOverride;
```
We think an enumeration would work best here with the options `default`, ~~`enabled`~~, and `disabled`. We are actually on the fence about including `enabled` as we have some fields that we intentionally disable to prevent Auth flows from breaking (such as disabling but displaying `username` throughout the signup flow). Developers should only be using `disabled` for fields that will be populated programmatically and `default` for everything else.

angelorodem replied:
Changed, please check if the new code helps,
though i still kept the enabled enum state, let me know if i should remove it, i think it is still important to have it.

tyllark commented on packages/authenticator/amplify_authenticator/lib/src/widgets/form_field.dart:
```diff
@@ -96,6 +98,20 @@ abstract class AuthenticatorFormField<
   final Iterable<String>? autofillHints;
 
+  /// Optional text controller exposed by text-driven form fields.
+  TextEditingController? get controller => null;
```
Additionally we should name this property `authenticatorTextFieldController` instead of `controller`. I know `TextEditingController` uses the name `controller`, but we think it is better to have a more precise name.  This change should als be made for all the FormField named constructors as well.

angelorodem commented:
@tyllark  I've addressed most of your feedback, including the API naming changes and the controller class structure.

Regarding the update cycle and complexity: I explored alternative approaches to avoid state updates during the build phase, but the current implementation (leveraging post-frame callbacks) proved to be the most robust way to handle bidirectional synchronization without race conditions. I wasn't able to simplify it further without reintroducing issues.

tyllark commented on packages/authenticator/amplify_authenticator/lib/src/mixins/authenticator_text_field.dart:
```diff
@@ -19,7 +19,8 @@ mixin AuthenticatorTextField<
 
   @protected
-  TextEditingController? get textController => null;
+  TextEditingController? get textController =>
```
This type should also be `AuthenticatorTextFieldController`

tyllark commented on packages/authenticator/amplify_authenticator/lib/src/enums/authenticator_text_enabled_override.dart:
```diff
@@ -0,0 +1,14 @@
+  defaultSetting,
+
+  /// Force the field to be enabled.
+  enabled,
```
Sorry if I wasn't clear with my previous comment. I think we should remove `enabled`. If the default setting causes is disabling the field it is because the field is not allowed to be modified at that time otherwise the Auth flow could break. Let me know if you have a use case that requires `enabled`.

tyllark replied:
Currently `confirm_sign_up_form_field.dart` is the only example I'm aware of  where we disable form fields.

angelorodem replied:
changed.

tyllark commented on packages/authenticator/amplify_authenticator/lib/src/mixins/authenticator_username_field.dart:
```diff
@@ -366,11 +398,14 @@ mixin AuthenticatorUsernameField<
-        enabled: enabled,
+        enabled: enabled
+            ? AuthenticatorTextEnabledOverride.enabled
+            : AuthenticatorTextEnabledOverride.disabled,
```
We should retain the original `AuthenticatorTextEnabledOverride` value rather than converting it to a `bool` and back. You might need to add a getter to `AuthenticatorFormFieldState`

tyllark commented on packages/authenticator/amplify_authenticator/lib/src/widgets/form_fields/confirm_sign_up_form_field.dart:
```diff
@@ -94,9 +94,14 @@ abstract class _ConfirmSignUpFormFieldState<FieldValue extends Object>
+        return false;
+      case AuthenticatorTextEnabledOverride.defaultSetting:
+      case null:
+        break;
```
I believe we can revert this change with the `effectiveEnabled` property I suggested.

angelorodem replied:
reverted.

angelorodem commented on packages/authenticator/amplify_authenticator/lib/src/mixins/authenticator_username_field.dart:
```diff
@@ -24,7 +24,8 @@ mixin AuthenticatorUsernameField<
 
   @protected
-  TextEditingController? get textController => null;
+  TextEditingController? get textController =>
```
fixed.

angelorodem commented on packages/authenticator/amplify_authenticator/lib/src/widgets/form_field.dart:
```diff
@@ -187,7 +196,17 @@ abstract class AuthenticatorFormFieldState<
+      case null:
+        return true;
+    }
+  }
```
changed.

angelorodem commented on packages/authenticator/amplify_authenticator/lib/src/widgets/form_fields/phone_number_field.dart:
```diff
@@ -77,10 +82,17 @@ class _AuthenticatorPhoneFieldState<FieldType extends Enum>
+      case null:
+        return super.enabled;
+    }
+  }
```
changed.

angelorodem commented:
@tyllark, i've applied the changes as you mentioned, please let me know if there is any other needed changes.

tyllark commented:
Hi @angelorodem, could you please update the branch (otherwise the commit will be associated with me). I've asked @ekjotmultani to take a quick glance, but it looks like we should be good to go!

angelorodem commented:
@tyllark Branch updated, let me know if there is anything else

angelorodem commented:
Hi @tyllark ,sorry for the bug, yes, feel free to chime in and finish, i would appreciate a lot your help, it is better to someone with better knowledge of the repo to finish this change.

Thank you!

tyllark commented:
@angelorodem I just restarted the failed tests. Some of our Github actions are a bit flakey and might fail due to Android emulators not starting up or network issues. These tests all passed before so I don't expect there to be any problems but if there are I'll push a fix.
